Change from LunrJs to MinisearchJs for client-side search

Hetarth02 opened the pull request:
This PR aims to close Issue #2141.

mortenpi commented:
Something I noticed while testing is that it likes to repeat page headings a lot:

![image](https://github.com/JuliaDocs/Documenter.jl/assets/147757/73cbd83c-101e-4017-be11-cd3f63ea96a7)

The lunr based approach doesn't do that:

![image](https://github.com/JuliaDocs/Documenter.jl/assets/147757/5f9ef669-4e35-45d5-8ca1-797b15b0b63c)

---

One other note: the PR has also a lot of non-substantial formatting changes. While I don't disagree with the changes, it's generally a good idea to make the formatting changes independently. In this case, though, it's a little less important -- we're kinda replacing the whole `search.js` anyway. But it is a little tricky to see right now how the implementations differ, and where they're the same.

Hetarth02 commented:
I have fixed the duplication issue.

Additionally, I have also updated the UI for the search results. The new UI looks like this,

![image](https://github.com/JuliaDocs/Documenter.jl/assets/80420115/820367f4-a122-443c-981b-3ebba19cb6a7)

![image](https://github.com/JuliaDocs/Documenter.jl/assets/80420115/508eed25-644d-4c9c-93a7-6f701dc392a6)

The new UI provides context(if possible) as to where the search matches, while also displaying the category badge on the right and title on the left with link at the bottom.

Hetarth02 commented:
I have fixed the duplication issue.

Additionally, I have updated the search result UI to provide better context for the results like displaying not only title but the matched string(if found) and the category badge too.

The new UI looks like this,

![image](https://github.com/JuliaDocs/Documenter.jl/assets/80420115/d77b3a2d-c85e-4bdb-833b-38cc79dc008c)

![image](https://github.com/JuliaDocs/Documenter.jl/assets/80420115/46e43ef9-42ff-4cbf-bbba-209fabaaa127)

mortenpi reviewed (changes requested):
Awesome! Mostly just a bit of bikeshedding. The main thing would be moving the search layout stuff into a shared SCSS file, to avoid duplication, but then I'd pretty much be happy to merge I think.

Hetarth02 commented:
Alright, will clean up the code as suggested.

Hetarth02 commented on assets/html/scss/documenter-light.scss:
```diff
@@ -40,4 +40,57 @@ code.language-julia-repl > span.hljs-meta {
+  background-color: hsl(48, 100%, 67%);
+}
+
+.search-divider {
```
Hmm, this won't on `a` tag because we are rounding off the borders so it will look like this.

![image](https://github.com/JuliaDocs/Documenter.jl/assets/80420115/4a3059ea-69c6-4de6-82a7-4eddcfdbab2a)

Two ways to solve this,

1. Remove the `border-radius` from `a` tag
2. Keep this as it is.

mortenpi replied:
Fair point. Let's keep it as is.

Hetarth02 commented on assets/html/scss/documenter-light.scss:
```diff
@@ -40,4 +40,57 @@ code.language-julia-repl > span.hljs-meta {
+}
+
+.search-result-highlight {
+  background-color: hsl(48, 100%, 67%);
```
This color suits well in both light and dark mode, so I will chose this instead of `#edff00`.

fredrikekre commented on assets/html/scss/documenter-dark.scss:
```diff
@@ -122,4 +122,58 @@ html.theme--#{$themename} {
+  }
+  
+  .search-result-highlight {
+    background-color: greenyellow;
```
Wouldn't the orange one from the light theme work in the dark theme too? Or perhaps a slightly darker orange

mortenpi commented on assets/html/scss/documenter-light.scss:
```diff
@@ -40,4 +40,4 @@ code.language-julia-repl > span.hljs-meta {
 // Workaround to compile in highlightjs theme, so that we could have different
 // themes for both
-@import "highlightjs/default"
+@import "highlightjs/default";
```
```suggestion
@import "highlightjs/default";

```

mortenpi replied:
Needs a newline at the end as well.

mortenpi commented on assets/html/scss/documenter/_variables.scss:
```diff
@@ -89,3 +89,15 @@ $documenter-docstring-header-padding: 0.5rem $documenter-container-left-padding;
+// Search Results variables
+$search-result-link-hover: rgba(0, 128, 128, 0.1) !default;
+
+$search-333: #333 !default;
+$search-f1f5f9: #f1f5f9 !default;
+$search-whitesmoke: whitesmoke !default;
```
I think the variables should have a bit more "semantic" names. What role do they fill?

mortenpi commented on assets/html/scss/documenter-dark.scss:
```diff
@@ -122,4 +122,13 @@ html.theme--#{$themename} {
     color: #f8f8f2;
   }
+  
+  .search-result-link:hover .property-search-result-badge, .search-result-link:focus .property-search-result-badge {
+    color: $search-333 !important;
+    background-color: $search-f1f5f9 !important;
+  }
+  
+  .search-result-title {
+    color: $search-whitesmoke !important;
+  }
```
I think we should just be able to do this by overriding the variables in `documenter-dark.scss`, not the whole selectors?

mortenpi commented on assets/html/scss/documenter/layout/_search.scss:
```diff
@@ -14,5 +14,59 @@
+    
+    .search-result-highlight {
+      background-color: $search-result-highlight;
+      color: black !important;
```
Are these `!important`s important? I think ideally we should avoid them, unless we really need them. But this can always be "fixed" later as well.

mortenpi commented on assets/html/scss/documenter/layout/_search.scss:
```diff
@@ -14,5 +14,59 @@
     .docs-highlight {
       background-color: yellow;
     }
```
Any of the old ones we can remove? Is e.g. `.docs-highlight` used anywhere?

mortenpi reviewed (approved):
This LGTM, thank you @Hetarth02! I need to get the Prettier workflow working properly, so I'll hold off on merging until that passes.
